compiler/src/lnntoamm/Types.ts

```diff
@@ -242,6 +246,162 @@ class Builtin extends Type {
   }
 }
 
+export class FunctionType extends Type {
+  params: Type[];
+  retTy: Type;
+
+  get ammName(): string {
+    throw new Error('Method not implemented.');
+  }
+
+  constructor(ast: LPNode, params: Type[], retTy: Type) {
+    super('<function>', ast);
+    this.params = params;
+    this.retTy = retTy;
+  }
+
+  static matrixSelect(fns: Fn[], args: Type[], scope: Scope): [Fn, Type][] {
+    const originalLength = fns.length;
+    // remove any fns that shouldn't apply
+    const callTy = new FunctionType(new NulLP(), args, Type.generate());
+    fns = fns.filter((fn) => fn.ty.compatibleWithConstraint(callTy, scope));
+    // if it's 0-arity then all we have to do is grab the retTy of the fn
+    if (args.length === 0) {
+      return fns.reduce(
+        (fns, fn) => [...fns, [fn, fn.retTy.instance()]],
+        new Array<[Fn, Type]>(),
+      );
+    }
+    // and now to generate the matrix
+    // every argument is a dimension within the matrix, but we're
+    // representing each dimension _d_ as an index in the matrix
+    const matrix: Array<Type[]> = args.map((arg) => {
+      return arg.fnselectOptions();
+    });
+    // TODO: this weight system feels like it can be inaccurate
+    // the weight of a particular function is computed by the sum
+    // of the indices in each dimension, with the highest sum
+    // having the greatest preference
+    const fnsByWeight = new Map<number, [Fn, Type][]>();
+    const indices = matrix.map(() => 0);
+    // keep it as for instead of while for debugging reasons
+    for (let i = 0; ; i++) {
+      const weight = indices.reduce((w, c) => w + c);
+      const argTys = matrix.map((options, ii) => options[indices[ii]]);
+      const fnsForWeight = fnsByWeight.get(weight) || [];
+      fnsForWeight.push(
+        ...fns.reduce((fns, fn) => {
+          const retTy = fn.resultTyFor(argTys, scope);
+          if (retTy === null) {
+            return fns;
+          } else {
+            return [...fns, [fn, retTy] as [Fn, Type]];
+          }
+        }, new Array<[Fn, Type]>()),
+      );
+      fnsByWeight.set(weight, fnsForWeight);
+      if (
+        indices.every(
+          (idxInDim, dimIdx) => idxInDim === matrix[dimIdx].length - 1,
+        )
+      ) {
+        break;
+      }
```

Comment:
Is this conditional guaranteed to always eventually execute? Otherwise this for loop can potentially infinite loop.

Reply:
yes. the index iteration is similar to how normal addition is performed. 

Say dimension 0 (using 0-indexing since that's how the `matrix` var works) has 2 elements, dimension 1 has 1 element, and dimension 2 has 2 elements. The first loop run would be `indices=[0,0,0]`. the `reduceRight` on the next line performs the addition-like iteration, resulting in `indices=[0,0,1]`. On the next loop, it's `indices=[1,0,0]` because of the carry skipping dimension 1. After this `if` check gets performed, it performs the final addition to get `indices=[1,0,1]` but must go to the beginning of the loop to calculate weight, etc before finally passing this `if` check and `break`ing